include/libcyphal/transport/can/transport.hpp

```diff
@@ -168,24 +203,143 @@ class TransportImpl final : public ICanTransport, private TransportDelegate
+        cetl::optional<AnyError> maybe_error{};
+
+        for (Media& media : media_array_)
+        {
+            media.canard_tx_queue.mtu_bytes = media.interface.getMtu();
+
+            const auto result =
+                ::canardTxPush(&media.canard_tx_queue, &canard_instance(), deadline, &metadata, payload_size, payload);
+            if (result < 0)
+            {
+                maybe_error = TransportDelegate::anyErrorFromCanard(result);
+            }
+        }
+
+        return maybe_error;
+    }
+
+    // MARK: Privates:
+
+    class Media final
+    {
+    public:
+        Media(const std::size_t _index, IMedia& _interface, const std::size_t tx_capacity)
+            : index{static_cast<std::uint8_t>(_index)}
+            , interface{_interface}
+            , canard_tx_queue{::canardTxInit(tx_capacity, _interface.getMtu())}
+        {
+        }
+        ~Media()                           = default;
+        Media(const Media&)                = delete;
+        Media(Media&&) noexcept            = default;
+        Media& operator=(const Media&)     = delete;
+        Media& operator=(Media&&) noexcept = delete;
+
+        const std::uint8_t index;
+        IMedia&            interface;
+        CanardTxQueue      canard_tx_queue;
+    };
+    using MediaArray = libcyphal::detail::VarArray<Media>;
+
+    template <typename T, typename Reducer>
+    CETL_NODISCARD T reduceMedia(const T init, Reducer reducer) const
+    {
+        T acc = init;
+        for (const Media& media : media_array_)
         {
-            CETL_DEBUG_ASSERT(media_array_[media_index] != nullptr, "Expected media interface.");
-            IMedia& media = *media_array_[media_index];
+            acc = reducer(std::forward<T>(acc), media);
+        }
+        return acc;
+    }
 
+    CETL_NODISCARD cetl::optional<AnyError> ensureNewSessionFor(const CanardTransferKind transfer_kind,
+                                                                const PortId             port_id,
+                                                                const PortId             max_port_id) noexcept
+    {
+        if (port_id > max_port_id)
+        {
+            return ArgumentError{};
+        }
+
+        const auto hasSubscription = ::canardRxHasSubscription(&canard_instance(), transfer_kind, port_id);
+        CETL_DEBUG_ASSERT(hasSubscription >= 0, "There is no way currently to get an error here.");
+        if (hasSubscription > 0)
+        {
+            return AlreadyExistsError{};
+        }
+
+        return {};
+    }
+
+    static MediaArray make_media_array(cetl::pmr::memory_resource& memory,
+                                       const std::size_t           tx_capacity,
+                                       const std::size_t           media_count,
+                                       const cetl::span<IMedia*>   media_interfaces)
+    {
+        MediaArray media_array{media_count, &memory};
+        media_array.reserve(media_count);
```

Comment:
@thirtytwobits Looks like actually I don't need to check it - b/c it is just a micro optimization so that below `media_array.emplace_back` won't try to **reallocate** array's memory. Also such `emplace_back` either throws c++ exception or just does nothing (when `--no-exceptions`). My point - there is no undefined behavior IMO; in worst case transport will end up with less media interfaces than it was requested (and, honestly, transport anyway will do no use if there is no memory at `mr_` even for such small array of media-s).

Reply:
This is not acceptable.

If there's no memory to store the media array, transport construction must fail with `MemoryError`. How about you create the array in the transport factory and then pass it to the ctor in case of success? This way you can easily handle OOM.

Reply:
@pavel-kirienko  Fixed, plz see the last commit.